File: src/main/java/com/dunemaster/unrolleddeque/UnrolledLinkedListDeque.java
package com.dunemaster.unrolleddeque;

import java.util.Collection;
import java.util.Deque;
import java.util.HashSet;
import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Set;

public class UnrolledLinkedListDeque<E> implements Deque<E> {

    private static final int DEFAULT_BLOCK_SIZE = 128;
    private final int blockSize;
    private final int center;
    private Node<E> head;
    private Node<E> tail;
    private int size;
    private int indexInHeadBlock;
    private int indexInTailBlock;

    private static final class Node<E> {

        private final E[] elements;
        private Node<E> next;
        private Node<E> prev;

        @SuppressWarnings("unchecked")
        Node(int blockSize) {
            elements = (E[]) new Object[blockSize];
        }

    }

    public UnrolledLinkedListDeque() {
        this(DEFAULT_BLOCK_SIZE);
    }

    public UnrolledLinkedListDeque(int blockSize) {
        this.blockSize = blockSize;
        this.center = blockSize / 2 - 1;
        validateBlockSize();
        head = new Node<>(blockSize);
        setToClearState();
    }

    private void validateBlockSize() {
        if (blockSize <= 0) throw new IllegalArgumentException("blockSize must be positive");
        if (blockSize % 2 != 0) throw new IllegalArgumentException("blockSize must be even");
    }

    private void setToClearState() {
        tail = head;
        size = 0;
        indexInHeadBlock = center + 1;
        indexInTailBlock = center;
    }

    @Override
    public void addFirst(E e) {
        if (!tryAddFirst(e))
            throw new IllegalStateException();
    }

    @Override
    public void addLast(E e) {
        tryAddLast(e);
    }

    @Override
    public boolean offerFirst(E e) {
        return tryAddFirst(e);
    }

    @Override
    public boolean offerLast(E e) {
        tryAddLast(e);
        return true;
    }

    @Override
    public E removeFirst() {
        throwIfEmpty();
        return tryRemoveFirst();
    }

    @Override
    public E removeLast() {
        throwIfEmpty();
        return tryRemoveLast();
    }

    @Override
    public E pollFirst() {
        return tryRemoveFirst();
    }

    @Override
    public E pollLast() {
        return tryRemoveLast();
    }

    @Override
    public E getFirst() {
        throwIfEmpty();
        return tryGetFirst();
    }

    @Override
    public E getLast() {
        throwIfEmpty();
        return tryGetLast();
    }

    @Override
    public E peekFirst() {
        return tryGetFirst();
    }

    @Override
    public E peekLast() {
        return tryGetLast();
    }

    @Override
    public boolean removeFirstOccurrence(Object o) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean removeLastOccurrence(Object o) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean add(E e) {
        return tryAddLast(e);
    }

    @Override
    public boolean offer(E e) {
        return offerLast(e);
    }

    @Override
    public E remove() {
        return removeFirst();
    }

    @Override
    public E poll() {
        return pollFirst();
    }

    @Override
    public E element() {
        return getFirst();
    }

    @Override
    public E peek() {
        return tryGetFirst();
    }

    @Override
    public boolean addAll(Collection<? extends E> elements) {
        boolean anyAdded = false;
        for (E element : elements) {
            boolean added = tryAddLast(element);
            anyAdded = anyAdded || added;
        }
        return anyAdded;
    }

    @Override
    public void push(E e) {
        addFirst(e);
    }

    @Override
    public E pop() {
        return removeFirst();
    }


    @Override
    public int size() {
        return size;
    }

    @Override
    public Iterator<E> iterator() {
        return new Iterator<>() {
            private Node<E> currentBlock = head;
            private int indexInCurrentBlock = indexInHeadBlock;
            private int remaining = size;

            @Override
            public boolean hasNext() {
                return remaining > 0;
            }

            @Override
            public E next() {
                if (!hasNext()) {
                    throw new NoSuchElementException();
                }
                E element = (E) currentBlock.elements[indexInCurrentBlock];
                indexInCurrentBlock++;
                remaining--;
                if (indexInCurrentBlock == blockSize && remaining > 0) {
                    currentBlock = currentBlock.next;
                    indexInCurrentBlock = 0;
                }
                return element;
            }
        };
    }

    @Override
    public E[] toArray() {
        //TODO
        return (E[])new Object[size];
    }

      @Override
    public <T> T[] toArray(T[] ts) {
        //TODO
        return null;
    }

    @Override
    public Iterator<E> descendingIterator() {
        return new Iterator<E>() {
            private Node<E> currentBlock = tail;
            private int indexInCurrentBlock = indexInTailBlock;
            private int remaining = size;

            @Override
            public boolean hasNext() {
                return remaining > 0;
            }

            @Override
            public E next() {
                if (!hasNext()) {
                    throw new NoSuchElementException();
                }

                E element = currentBlock.elements[indexInCurrentBlock];
                indexInCurrentBlock--;
                remaining--;
                if (indexInCurrentBlock < 0 && remaining > 0) {
                    currentBlock = currentBlock.prev;
                    indexInCurrentBlock = blockSize - 1;
                }
                return element;
            }
        };
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }


    @Override
    public void clear() {
        head = new Node<>(blockSize);
        setToClearState();
    }

    private E tryGetFirst() {
        if (size == 0) {
            return null;
        }
        return (E) head.elements[indexInHeadBlock];
    }

    private E tryGetLast() {
        if (size == 0) {
            return null;
        }
        return (E) tail.elements[indexInTailBlock];
    }

    private boolean tryAddFirst(E element) {
        indexInHeadBlock--;
        if (indexInHeadBlock < 0) {
            Node<E> newNode = new Node<>(blockSize);
            if (size != 0) { // Only link nodes if the deque is not empty
                newNode.next = head;
                head.prev = newNode;
            }
            head = newNode;
            indexInHeadBlock = blockSize - 1;
        }
        head.elements[indexInHeadBlock] = element;
        size++;
        return true;
    }

    private boolean tryAddLast(E element) {
        indexInTailBlock++;
        if (indexInTailBlock == blockSize) {
            Node<E> newNode = new Node<>(blockSize);
            tail.next = newNode;
            newNode.prev = tail;
            tail = newNode;
            indexInTailBlock = 0;
        }
        tail.elements[indexInTailBlock] = element;
        size++;
        return true;
    }

    private E tryRemoveFirst() {
        if (size == 0) {
            return null;
        }
        // Adjust for the circular nature of the deque
        int effectiveIndex = indexInHeadBlock;
        E element = (E) head.elements[effectiveIndex];
        head.elements[effectiveIndex] = null; // Help GC
        indexInHeadBlock++;
        if (indexInHeadBlock == blockSize) {
            if (head != tail) {
                head = head.next;
                head.prev = null;
            }
            indexInHeadBlock = 0;
        }
        size--;
        if (size == 0) {
            setToClearState();
        }
        return element;
    }

    private E tryRemoveLast() {
        if (size == 0) {
            return null;
        }

        E element = (E) tail.elements[indexInTailBlock];
        // releasing memory!
        tail.elements[indexInTailBlock] = null;
        indexInTailBlock--;
        size--;
        if (indexInTailBlock < 0) {
            // remove the last node
            if (head == tail) {
                if (size == 0) {
                    setToClearState();
                }
            } else {
                Node<E> prev = tail.prev;
                prev.next = null;
                tail = tail.prev;
                indexInTailBlock = blockSize - 1;
            }
        }

        return element;
    }

    private void throwIfEmpty() {
        if (size == 0)
            throw new NoSuchElementException();
    }

    @Override
    public boolean removeAll(Collection<?> collection) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean retainAll(Collection<?> collection) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean remove(Object o) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean containsAll(Collection<?> elements) {
        if (elements.isEmpty()) {
            return true;
        }
        if (elements.size() == 1) {
            return contains(elements.iterator().next());
        }
        Set<Object> testSet = elements instanceof Set ? (Set)elements : new HashSet<>(elements);
        for (Object dequeElement : this) {
            if (testSet.contains(dequeElement)) {
                testSet.remove(dequeElement);
            }
        }
        return testSet.isEmpty();
    }

    @Override
    public boolean contains(Object o) {
        for (Object object : this) {
            if (object.equals(o)) {
                return true;
            }
        }
        return false;
    }

}

File: src/main/kotlin/com/dunemaster/unrolleddeque/UnrolledLinkedListDeque.kt
package com.dunemaster.unrolleddeque


const val DEFAULT_BLOCK_SIZE : Int = 128

/**
 * This is an implementation of an unrolled linked list deque, where data is stored in blocks,
 * which gives better cache locality and fewer allocations than a regular linked list deque.
 *
 * The idea of having two indexes, one for the left and one for the right, is to allow
 * for efficient insertion and removal at both ends of the list. It is borrowed from the Python
 * implementation of deque.
 *
 * Currently, capacity restrictions are not supported
 */
class UnrolledLinkedListDeque<E>(
    private val blockSize : Int  = DEFAULT_BLOCK_SIZE,
    private val center : Int = blockSize / 2 - 1,
                          ) :  java.util.Deque<E> {

    private var head: Node<E>

    private lateinit var tail: Node<E>

    init {
        require(blockSize > 0) { "blockSize must be positive" }
        require(blockSize % 2 == 0) { "blockSize must be even" }
        head = Node(blockSize)
        setToClearState()
    }

    constructor(blockSize: Int) : this(blockSize, blockSize / 2 - 1) {

    }

    private var indexInHeadBlock : Int = center + 1
    private var indexInTailBlock : Int = center

    private class Node<E>(blockSize: Int) {
        val elements = arrayOfNulls<Any>(blockSize)

        var prev: Node<E>? = null

        var next: Node<E>? = null
    }



    override var size: Int = 0
        private set

    override fun addAll(elements: Collection<E>): Boolean {
        var anyAdded = false
        for (element in elements) {
            val added = tryAddLast(element)
            anyAdded = anyAdded || added
        }
        return anyAdded
    }

    override fun clear() {
        head = Node(blockSize)
        setToClearState()
    }

    override fun iterator(): MutableIterator<E> {
       val it =  object : MutableIterator<E> {
                private var currentBlock = head
                private var indexInCurrentBlock = indexInHeadBlock
                private var remaining = size

                override fun hasNext(): Boolean {
                    return remaining > 0
                }

                override fun next(): E {
                    if (remaining == 0) {
                        throw NoSuchElementException()
                    }
                    val element = currentBlock.elements[indexInCurrentBlock] as E
                    indexInCurrentBlock++
                    remaining--
                    if (indexInCurrentBlock == blockSize && remaining > 0) {
                        currentBlock = currentBlock.next!!
                        indexInCurrentBlock = 0
                    }

                    return element
                }

           override fun remove() {
                throw UnsupportedOperationException()
           }
       }
        return  it
    }

    override fun descendingIterator(): MutableIterator<E> {
        val it = object : MutableIterator<E> {
            private var currentBlock = tail
            private var indexInCurrentBlock = indexInTailBlock
            private var remaining = size

            override fun hasNext(): Boolean {
                return remaining > 0
            }

            override fun next(): E {
                if (remaining == 0) {
                    throw NoSuchElementException()
                }
                val element = currentBlock.elements[indexInCurrentBlock] as E
                indexInCurrentBlock--
                remaining--
                if (indexInCurrentBlock < 0 && remaining > 0) {
                    currentBlock = currentBlock.prev!!
                    indexInCurrentBlock = blockSize - 1
                }

                return element
            }

            override fun remove() {
                throw UnsupportedOperationException()
            }
        }
        return it
    }

    override fun remove(): E = removeFirst()

    override fun isEmpty(): Boolean {
        return size == 0
    }

    override fun containsAll(elements: Collection<E>): Boolean {
        if (elements.isEmpty()) {
            return true
        }
        if (elements.size == 1) {
            return contains(elements.first())
        }
        val testSet = HashSet(elements)
        for (dequeElement in this) {
          if (testSet.contains(dequeElement)) {
              testSet.remove(dequeElement)
          }
        }
        return testSet.isEmpty()
    }

    override fun contains(element: E): Boolean {
       return indexOf(element) != -1
    }

    override fun poll(): E? = pollFirst()

    override fun element(): E  = getFirst()

    override fun peek(): E? =  peekFirst()

    override fun removeFirst(): E {
        throwIfEmpty()
        return tryRemoveFirst()!!
    }

    override fun removeLast(): E {
        throwIfEmpty()
        return tryRemoveLast()!!
    }

    override fun pollFirst(): E? = tryRemoveFirst()

    override fun pollLast(): E? {
        return tryRemoveLast()
    }

    override fun getFirst(): E {
        throwIfEmpty()
        return tryGetFirst()!!
    }

    override fun getLast(): E {
        return tryGetLast() ?: throw NoSuchElementException()
    }

    override fun peekFirst(): E?  =  tryGetFirst()

    override fun peekLast(): E? {
        return tryGetLast()
    }

    override fun pop(): E = removeFirst()


    override fun push(element: E) = addFirst(element)

    override fun offerLast(element: E): Boolean {
        tryAddLast(element)
        return true
    }

    override fun offerFirst(element: E): Boolean {
        return tryAddFirst(element)
    }

    override fun addLast(element: E) {
        tryAddLast(element)
    }

    override fun addFirst(element: E) {
       if (!tryAddFirst(element))
           throw IllegalStateException()
    }

    override fun offer(element: E): Boolean {
        tryAddLast(element)
        return true
    }
    override fun add(element: E) = tryAddLast(element)

    override fun retainAll(elements: Collection<E>): Boolean {
        throw UnsupportedOperationException()
    }

    override fun removeAll(elements: Collection<E>): Boolean {
        throw UnsupportedOperationException()
    }

    override fun remove(element: E): Boolean {
        throw UnsupportedOperationException()
    }

    override fun removeFirstOccurrence(element: Any?): Boolean {
        throw UnsupportedOperationException()
    }

    override fun removeLastOccurrence(element: Any?): Boolean {
        throw UnsupportedOperationException()
    }

    private fun tryAddLast(element: E): Boolean {
        indexInTailBlock++
        if (indexInTailBlock == blockSize) {
            val newNode = Node<E>(blockSize)
            tail.next = newNode
            newNode.prev = tail
            tail = newNode
            indexInTailBlock = 0
        }
        tail.elements[indexInTailBlock] = element
        size++
        return true
    }

    private fun tryGetLast() : E? {
        if (size == 0) {
            return null
        }
        return tail.elements.get(indexInTailBlock) as E
    }

    private fun tryRemoveLast(): E? {
        if (size == 0) {
            return null
        }
        val element = tail.elements[indexInTailBlock]
        // releasing memory!
        tail.elements[indexInTailBlock] = null
        indexInTailBlock--
        size--
        if (indexInTailBlock < 0) {
            // remove the last node
            if (head == tail) {
                if (size == 0) {
                    setToClearState()
                }
            } else {
                val prev = tail.prev
                prev!!.next = null
                tail = tail.prev!!
                indexInTailBlock = blockSize - 1
            }
        }
        return element as E

    }

    private fun tryAddFirst(element: E): Boolean {
        indexInHeadBlock--
        if (size != 0) {
            if (indexInHeadBlock < 0) {
                val newNode = Node<E>(blockSize)
                head.prev = newNode
                newNode.next = head
                head = newNode
                indexInHeadBlock = blockSize - 1
            }
        }
        head.elements[indexInHeadBlock] = element
        size++
        return true
    }

    private fun tryGetFirst(): E? {
        if (size == 0) {
            return null
        }
        return head.elements[indexInHeadBlock] as E
    }

    private fun tryRemoveFirst(): E? {
        if (size == 0) {
            return null
        }
        val element = head.elements[indexInHeadBlock]
        // releasing memory!
        head.elements[indexInHeadBlock] = null
        indexInHeadBlock++
        size--
        if (indexInHeadBlock == blockSize) {
            // remove the first node
            if (head == tail) {
                if (size == 0) {
                    setToClearState()
                }
            } else {
                val next = head.next
                next!!.prev = null
                head = next
                indexInHeadBlock = 0
            }
        }
        return element as E
    }

    private inline fun throwIfEmpty() {
        if (size == 0) {
            throw NoSuchElementException()
        }
    }

    private fun setToClearState() {

        tail = head
        size = 0

        indexInHeadBlock = center + 1
        indexInTailBlock = center
    }

}